lib/robots/dor_repo/gis_assembly/generate_mods.rb

```diff
@@ -10,6 +10,69 @@ def initialize
           super('gisAssemblyWF', 'generate-mods', check_queued_status: true) # init LyberCore::Robot
         end
 
+        # `perform` is the main entry point for the robot. This is where
+        # all of the robot's work is done.
+        #
+        # @param [String] druid -- the Druid identifier for the object to process
+        def perform(druid)
+          druid = druid.delete_prefix('druid:')
+          LyberCore::Log.debug "generate-mods working on #{druid}"
+
+          rootdir = GisRobotSuite.locate_druid_path druid, type: :stage
+
+          # short-circuit if already have MODS file
+          desc_metadata = File.join(rootdir, 'metadata', 'descMetadata.xml')
+          if File.size?(desc_metadata)
+            LyberCore::Log.info "generate-mods: #{druid} found existing #{desc_metadata}"
+            return
+          end
+
+          geo_metadata_file = File.join(rootdir, 'metadata', 'geoMetadata.xml')
+          raise "generate-mods: #{druid} cannot locate #{geo_metadata_file}" unless File.size?(geo_metadata_file)
+
+          # parse geometadata as input to MODS transform
+          geo_metadata_rdf_xml = Nokogiri::XML(File.read(geo_metadata_file))
+
+          # detect fileFormat and geometryType
+          shp_file = Dir.glob("#{rootdir}/temp/*.shp").first
+          if shp_file.nil?
+            geometryType = 'Raster'
+            tif_file = Dir.glob("#{rootdir}/temp/*.tif").first
+            if tif_file.nil?
+              metadata_xml_file = Dir.glob("#{rootdir}/temp/*/metadata.xml").first
+              if metadata_xml_file.nil?
+                raise "generate-mods: #{druid} cannot detect fileFormat: #{rootdir}/temp"
+              else
+                fileFormat = 'ArcGRID'
+              end
+            else
+              fileFormat = 'GeoTIFF'
+            end
+          else
+            geometryType = geometry_type_ogrinfo(shp_file)
+            geometryType = 'LineString' if geometryType =~ /^Line/
+            fileFormat = 'Shapefile'
+          end
+
+          # load PURL
+          purl = Settings.purl.url + "/#{druid.gsub(/^druid:/, '')}"
+
+          # clean up geo_metadata_rdf_xml to not generate transforms
+          mods_xml_file = File.join(rootdir, 'metadata', 'descMetadata.xml')
+          File.open(mods_xml_file, 'wb') do |file_content|
+            file_content << to_mods(geo_metadata_rdf_xml,
+                                    geometryType: geometryType,
+                                    fileFormat: fileFormat,
+                                    purl: purl).to_xml(index: 2)
+          rescue ArgumentError => e
+            raise "generate-mods: #{druid} cannot process MODS: #{e}"
+          end
+
+          raise "generate-mods: #{druid} did not write MODS correctly" unless File.size?(mods_xml_file)
+        end
+
+        private
```

Comment:
this looks like more change than it is -- I hardly changed this at all and made the private methods ... private.  That is in a separate commit, if you want to see the changes before that happened.